File: backend/src/services/chatService.ts
import type { ChatRequest, ChatResponse } from "@agripulse/shared";
import { chatRepository } from "../repositories/chatRepository.js";
import { commodityRepository } from "../repositories/commodityRepository.js";
import { stateMetadata } from "../utils/stateMetadata.js";
import { intelligenceClient } from "./intelligenceClient.js";
import { priceService } from "./priceService.js";
import { productionService } from "./productionService.js";
import { stateService } from "./stateService.js";

function detectState(message: string, explicit?: string): string | undefined {
  const normalized = message.toLowerCase();
  for (const [code, state] of Object.entries(stateMetadata)) {
    if (normalized.includes(state.name.toLowerCase()) || new RegExp(`\\b${code}\\b`).test(message)) return code;
  }
  return explicit?.toUpperCase();
}

export class ChatService {
  health(): Promise<{ status: string; model_status: string; model_provider: string }> {
    return intelligenceClient.health();
  }

  async answer(input: ChatRequest): Promise<ChatResponse> {
    const recentMessages = await chatRepository.findRecentContext(input.sessionId);
    const priorMetadata = [...recentMessages].reverse().find((message) => message.metadata)?.metadata as
      | { stateCode?: string | null; commoditySlug?: string | null }
      | undefined;
    const normalized = input.message.toLowerCase();
    const asksAcrossStates = /\b(which|what|top|leading)\s+states?\b|\bstates?\s+(lead|produce|rank)\b|\bnational(?:ly)?\b/.test(normalized);
    const asksForStateMix = /\b(compare|list|show)\b.*\b(crops|commodities|agricultural mix)\b|\b(dominant|leading|top)\s+crop\b|\bcrop mix\b/.test(normalized);
    const commodities = await commodityRepository.findAll();
    const mentionedCommodity = commodities.find((commodity) =>
      normalized.includes(commodity.name.toLowerCase()) ||
      normalized.includes(commodity.slug) ||
      normalized.includes(commodity.name.toLowerCase().replace(/s$/, "")) ||
      normalized.includes(commodity.slug.replace(/s$/, ""))
    );
    const selectedCommodity = commodities.find((commodity) => commodity.slug === input.commoditySlug);
    const mentionedState = detectState(input.message);
    const stateCode = mentionedState ?? (asksAcrossStates ? undefined : input.stateCode ?? priorMetadata?.stateCode ?? undefined);
    const resolvedCommodity = mentionedCommodity ?? (asksForStateMix ? undefined : selectedCommodity ?? commodities.find((commodity) => commodity.slug === priorMetadata?.commoditySlug));

    let response: ChatResponse;
    try {
      response = await intelligenceClient.answerChat({
        message: input.message,
        ...(input.sessionId ? { sessionId: input.sessionId } : {}),
        ...(resolvedCommodity ? { commoditySlug: resolvedCommodity.slug } : {}),
        ...(stateCode ? { stateCode } : {}),
        history: recentMessages.map((message) => ({
          role: message.role === "USER" ? "user" as const : "assistant" as const,
          content: message.content
        }))
      });
    } catch {
      const fallback = await this.databaseFallback(resolvedCommodity?.slug, stateCode);
      response = {
        ...fallback,
        evaluation: {
          confidenceScore: 90,
          groundednessScore: 100,
          citationCoverageScore: 100,
          relevanceScore: 85,
          status: "high",
          flags: ["deterministic_fallback", "intelligence_service_unavailable"]
        }
      };
    }

    const session = await chatRepository.createExchange(input.message, response.answer, {
      stateCode: stateCode ?? null,
      commoditySlug: resolvedCommodity?.slug ?? null,
      generatedBy: response.generatedBy,
      asOf: response.asOf
    }, input.sessionId);
    return { ...response, sessionId: session.id };
  }

  private async databaseFallback(commoditySlug?: string, stateCode?: string): Promise<ChatResponse> {
    if (stateCode) {
      const detail = await stateService.getDetail(stateCode);
      const requested = commoditySlug
        ? detail.production.find((item) => item.commoditySlug === commoditySlug)
        : undefined;
      if (commoditySlug && requested) {
        const valueText = requested.totalValueUsd === null
          ? "a stored production value is not available"
          : `the state production value is $${requested.totalValueUsd.toLocaleString()}`;
        return {
          answer: `${detail.stateName} produced ${requested.value.toLocaleString()} ${requested.unit} of ${requested.commodityName} in ${requested.year}; ${valueText}.${requested.unitPriceUsd === null ? "" : ` The implied unit value is $${requested.unitPriceUsd.toFixed(2)} per ${requested.unit.toLowerCase()}.`} This answer uses USDA NASS records stored in AgriPulse.`,
          model: "database-fallback",
          generatedBy: "fallback",
          citations: [{
            label: `${detail.stateName} ${requested.commodityName}`,
            value: `${requested.value.toLocaleString()} ${requested.unit} (${requested.year})`
          }],
          asOf: String(requested.year)
        };
      }
      if (commoditySlug && !requested) {
        const commodity = await commodityRepository.findByIdOrSlug(commoditySlug);
        return {
          answer: `No stored USDA NASS ${commodity?.name ?? commoditySlug} production record is available for ${detail.stateName}.`,
          model: "database-fallback",
          generatedBy: "fallback",
          citations: [],
          asOf: String(detail.year)
        };
      }
      const top = detail.topCommodity;
      const remaining = detail.production.slice(1, 4);
      const otherText = remaining.length
        ? ` Other tracked crops include ${remaining.map((item) => `${item.commodityName} (${item.value.toLocaleString()} ${item.unit})`).join(", ")}.`
        : "";
      return {
          answer: `${detail.stateName}'s leading stored crop is ${top.commodityName}: ${top.value.toLocaleString()} ${top.unit} in ${top.year}.${otherText}`,
        model: "database-fallback",
        generatedBy: "fallback",
        citations: detail.production.slice(0, 4).map((item) => ({
          label: `${detail.stateName} ${item.commodityName}`,
          value: `${item.value.toLocaleString()} ${item.unit} (${item.year})`
        })),
        asOf: String(detail.year)
      };
    }

    if (commoditySlug) {
      const commodity = await commodityRepository.findByIdOrSlug(commoditySlug);
      if (commodity) {
        const [latestPrice, production] = await Promise.all([
          priceService.getLatestForCommodity(commodity.slug),
          productionService.getMapData(commodity.slug)
        ]);
        const top = production.slice(0, 3);
        const priceText = latestPrice
          ? `${commodity.name}'s latest stored national price is ${latestPrice.value.toFixed(2)} ${latestPrice.unit}, dated ${latestPrice.priceDate.slice(0, 10)}.`
          : `No stored ${commodity.name} price is available.`;
        const productionText = top.length
          ? ` The leading stored production states are ${top.map((row) => `${row.stateName} (${row.value.toLocaleString()} ${row.unit})`).join(", ")}.`
          : " No production records are available.";
        return {
          answer: `${priceText}${productionText}`,
          model: "database-fallback",
          generatedBy: "fallback",
          citations: [
            ...(latestPrice ? [{ label: `${commodity.name} price`, value: `${latestPrice.value.toFixed(2)} ${latestPrice.unit} (${latestPrice.priceDate.slice(0, 10)})` }] : []),
            ...top.map((row) => ({ label: `${row.stateName} production`, value: `${row.value.toLocaleString()} ${row.unit} (${row.year})` }))
          ],
          asOf: latestPrice?.priceDate ?? String(top[0]?.year ?? new Date().getUTCFullYear())
        };
      }
    }

    const prices = await priceService.getLatest(5);
    if (!prices.length) {
      return {
        answer: "No agricultural records are available. Seed or synchronise the database first.",
        model: "database-fallback",
        generatedBy: "fallback",
        citations: [],
        asOf: new Date().toISOString()
      };
    }
    return {
      answer: `Latest stored national observations: ${prices.map((item) => `${item.commodityName} ${item.value.toFixed(2)} ${item.unit}`).join("; ")}. Check each observation date before treating it as current market information.`,
      model: "database-fallback",
      generatedBy: "fallback",
      citations: prices.map((item) => ({
        label: `${item.commodityName} price`,
        value: `${item.value.toFixed(2)} ${item.unit} (${item.priceDate.slice(0, 10)})`
      })),
      asOf: prices.reduce((latest, item) => item.priceDate > latest ? item.priceDate : latest, prices[0]!.priceDate)
    };
  }
}

export const chatService = new ChatService();

File: intelligence/app/chat_service.py
import json
import asyncio
from copy import deepcopy
from datetime import datetime, timezone
import httpx
from .config import get_settings
from .chat_evaluator import evaluate_answer
from .models import ChatCitation, ChatRequest, ChatResponse
from .repositories import agriculture_repository


STATE_NAMES = {
    "AL": "Alabama", "AK": "Alaska", "AZ": "Arizona", "AR": "Arkansas", "CA": "California",
    "CO": "Colorado", "CT": "Connecticut", "DE": "Delaware", "FL": "Florida", "GA": "Georgia",
    "HI": "Hawaii", "ID": "Idaho", "IL": "Illinois", "IN": "Indiana", "IA": "Iowa",
    "KS": "Kansas", "KY": "Kentucky", "LA": "Louisiana", "ME": "Maine", "MD": "Maryland",
    "MA": "Massachusetts", "MI": "Michigan", "MN": "Minnesota", "MS": "Mississippi", "MO": "Missouri",
    "MT": "Montana", "NE": "Nebraska", "NV": "Nevada", "NH": "New Hampshire", "NJ": "New Jersey",
    "NM": "New Mexico", "NY": "New York", "NC": "North Carolina", "ND": "North Dakota", "OH": "Ohio",
    "OK": "Oklahoma", "OR": "Oregon", "PA": "Pennsylvania", "RI": "Rhode Island", "SC": "South Carolina",
    "SD": "South Dakota", "TN": "Tennessee", "TX": "Texas", "UT": "Utah", "VT": "Vermont",
    "VA": "Virginia", "WA": "Washington", "WV": "West Virginia", "WI": "Wisconsin", "WY": "Wyoming",
}


def detect_state(message: str, explicit: str | None) -> str | None:
    if explicit:
        code = explicit.upper()
        return code if code in STATE_NAMES else None
    lowered = message.lower()
    for code, name in STATE_NAMES.items():
        if name.lower() in lowered or f" {code.lower()} " in f" {lowered} ":
            return code
    return None


def detect_states(message: str, explicit: str | None) -> list[str]:
    lowered = f" {message.lower()} "
    original = f" {message} "
    detected = [
        code for code, name in STATE_NAMES.items()
        if name.lower() in lowered or f" {code} " in original
    ]
    if explicit and explicit.upper() in STATE_NAMES and explicit.upper() not in detected:
        detected.append(explicit.upper())
    return detected


async def build_context(request: ChatRequest) -> tuple[dict, list[ChatCitation], str]:
    commodities, nationwide_production, all_prices = await asyncio.gather(
        agriculture_repository.list_commodities(),
        agriculture_repository.nationwide_latest_production(),
        agriculture_repository.all_latest_prices(),
    )
    lowered = request.message.lower()
    commodity = next(
        (
            item for item in commodities
            if item.slug == request.commodity_slug
            or item.name.lower() in lowered
            or item.slug in lowered
            or item.name.lower().removesuffix("s") in lowered
            or item.slug.removesuffix("s") in lowered
        ),
        None,
    )
    state_code = detect_state(request.message, request.state_code)
    state_codes = detect_states(request.message, request.state_code)
    relevant_nationwide_production = nationwide_production
    if commodity:
        relevant_nationwide_production = [
            row for row in nationwide_production if row.get("slug") == commodity.slug
        ]
    elif state_codes:
        relevant_nationwide_production = [
            row for row in nationwide_production if row.get("stateCode") in state_codes
        ]
    context: dict = {
        "question": request.message,
        "conversation": [turn.model_dump() for turn in request.history[-8:]],
        "state": None,
        "states": [],
        "commodity": None,
        "featuredPrices": [],
        "nationwideDataset": {
            "scope": "Question-relevant records retrieved from the complete nationwide USDA dataset; the selected state is a focus, not a database filter.",
            "production": relevant_nationwide_production,
            "latestNationalPrices": all_prices,
        },
    }
    citations: list[ChatCitation] = []
    as_of = datetime.now(timezone.utc).isoformat()

    if commodity:
        price = await agriculture_repository.national_price(commodity.slug)
        top_states = await agriculture_repository.top_states(commodity.slug)
        context["commodity"] = {"name": commodity.name, "slug": commodity.slug, "latestPrice": price, "topStates": top_states}
        if price:
            as_of = str(price["priceDate"])
            citations.append(ChatCitation(label=f"{commodity.name} price", value=f'{price["value"]:.2f} {price["unit"]} ({str(price["priceDate"])[:10]})'))
        for row in top_states[:3]:
            citations.append(ChatCitation(label=f'{row["stateCode"]} production', value=f'{row["value"]:,.0f} {row["unit"]} ({row["year"]})'))

    if state_code:
        production = await agriculture_repository.state_production(state_code)
        context["state"] = {"code": state_code, "name": STATE_NAMES[state_code], "production": production}
        for row in production[:3]:
            citations.append(ChatCitation(label=f'{STATE_NAMES[state_code]} {row["name"]}', value=f'{row["value"]:,.0f} {row["unit"]} ({row["year"]})'))

    if state_codes:
        state_rows = await asyncio.gather(*(agriculture_repository.state_production(code) for code in state_codes))
        context["states"] = [
            {"code": code, "name": STATE_NAMES[code], "production": rows}
            for code, rows in zip(state_codes, state_rows)
        ]
        for focused_state in context["states"]:
            requested = next(
                (row for row in focused_state["production"] if commodity and row.get("slug") == commodity.slug),
                None,
            )
            if requested:
                citations.append(ChatCitation(
                    label=f'{focused_state["name"]} {requested["name"]}',
                    value=f'{requested["value"]:,.0f} {requested["unit"]} ({requested["year"]})',
                ))

    if not commodity and not state_code:
        prices = await agriculture_repository.featured_prices()
        context["featuredPrices"] = prices
        for row in prices[:5]:
            citations.append(ChatCitation(label=f'{row["name"]} price', value=f'{row["value"]:.2f} {row["unit"]} ({str(row["priceDate"])[:10]})'))
        if prices:
            as_of = max(str(row["priceDate"]) for row in prices)

    context["rules"] = [
        "Use only this database context.",
        "The nationwideDataset is available for every question. Do not limit an answer to the focused state unless the question asks for that state only.",
        "Use nationwideDataset.production to compare, rank, aggregate, or answer questions about states other than the focused state.",
        "Always include dates, years, and units when quoting numbers.",
        "Do not describe periodic USDA observations as exchange-level real-time quotes.",
        "Say when the database does not contain enough evidence.",
        "Resolve pronouns and follow-up questions from the supplied conversation, but use database context for factual claims."
    ]
    return context, citations[:8], as_of


def fallback_answer(context: dict) -> str:
    state = context.get("state")
    states = context.get("states", [])
    commodity = context.get("commodity")
    sentences: list[str] = []
    if len(states) > 1 and commodity:
        comparisons = []
        for focused_state in states:
            requested = next(
                (row for row in focused_state.get("production", []) if row.get("slug") == commodity.get("slug")),
                None,
            )
            if requested:
                comparisons.append(
                    f'{focused_state["name"]} produced {requested["value"]:,.0f} {requested["unit"]} in {requested["year"]}'
                )
        if comparisons:
            return "; ".join(comparisons) + "."
    if state and state.get("production"):
        rows = state["production"]
        requested = next(
            (row for row in rows if commodity and row.get("slug") == commodity.get("slug")),
            None,
        )
        if requested:
            sentence = (
                f'{state["name"]} produced {requested["value"]:,.0f} {requested["unit"]} '
                f'of {requested["name"]} in {requested["year"]}'
            )
            if requested.get("unitPriceUsd") is not None:
                price_unit = requested["unit"].lower().removesuffix("s")
                sentence += f', with an implied value of ${requested["unitPriceUsd"]:,.2f} per {price_unit}'
            if requested.get("totalValueUsd") is not None:
                sentence += f' and total state production value of ${requested["totalValueUsd"]:,.0f}'
            sentences.append(sentence + ".")
            return " ".join(sentences)
        if commodity:
            return f'No stored USDA NASS {commodity["name"]} production record is available for {state["name"]}.'
        top = rows[0]
        sentences.append(f'{state["name"]}’s leading stored crop is {top["name"]}, with {top["value"]:,.0f} {top["unit"]} in {top["year"]}.')
        if len(rows) > 1:
            rest = ", ".join(f'{row["name"]} ({row["value"]:,.0f} {row["unit"]})' for row in rows[1:4])
            sentences.append(f"Other tracked crops are {rest}.")
    if commodity:
        price = commodity.get("latestPrice")
        if price:
            sentences.append(f'{commodity["name"]}’s latest stored national price is {price["value"]:.2f} {price["unit"]}, dated {str(price["priceDate"])[:10]}.')
        top_states = commodity.get("topStates", [])
        if top_states:
            sentences.append("The leading stored production states are " + ", ".join(f'{row["stateCode"]} ({row["value"]:,.0f} {row["unit"]})' for row in top_states[:3]) + ".")
    if not sentences and context.get("featuredPrices"):
        sentences.append("Latest stored prices: " + "; ".join(f'{row["name"]} {row["value"]:.2f} {row["unit"]}' for row in context["featuredPrices"][:5]) + ".")
    if not sentences:
        return "The AgriPulse database does not contain enough records to answer that question yet."
    return " ".join(sentences)


def response_text(payload: dict) -> str:
    for output in payload.get("output", []):
        if output.get("type") != "message":
            continue
        for content in output.get("content", []):
            if content.get("type") == "output_text" and content.get("text"):
                return str(content["text"]).strip()
    return ""


def model_context(context: dict) -> dict:
    formatted = deepcopy(context)

    def format_numbers(value):
        if isinstance(value, dict):
            return {
                key: (
                    f"{item:,.0f}" if key in {"value", "totalValueUsd"} and isinstance(item, (int, float))
                    else f"{item:,.2f}" if key == "unitPriceUsd" and isinstance(item, (int, float))
                    else format_numbers(item)
                )
                for key, item in value.items()
            }
        if isinstance(value, list):
            return [format_numbers(item) for item in value]
        return value

    return format_numbers(formatted)


async def answer_chat(request: ChatRequest) -> ChatResponse:
    context, citations, as_of = await build_context(request)
    prompt_context = model_context(context)
    settings = get_settings()
    system_prompt = (
        "You are AgriPulse Analyst. Give a clean, direct, concise agricultural-data answer using only the supplied database context. "
        "Never invent, estimate, round, or recalculate figures, and never make farm-specific claims from state-level data. "
        "Copy every number, date, year, unit, state code, and ranking exactly from the supplied JSON. "
        "Return only requested results that have records. Omit states, crops, metrics, and comparisons with no records. "
        "Do not provide suggestions, possible next steps, generic explanations, or repeat the question. "
        "If no requested result has a record, say so in one short sentence."
    )
    if len(context.get("states", [])) > 1 and context.get("commodity"):
        answer = fallback_answer(context)
        result = evaluate_answer(request.message, answer, context, citations, "fallback")
        return ChatResponse(
            answer=answer,
            model="agripulse-local",
            generated_by="fallback",
            citations=citations,
            as_of=as_of,
            evaluation=result.evaluation,
        )
    try:
        if settings.chat_provider.lower() != "openai":
            raise ValueError("OpenAI is not the configured primary provider")
        if not settings.openai_api_key:
            raise ValueError("OPENAI_API_KEY is not configured")
        async with httpx.AsyncClient(timeout=45.0) as client:
            response = await client.post(
                f"{settings.openai_base_url}/responses",
                headers={"Authorization": f"Bearer {settings.openai_api_key}"},
                json={
                    "model": settings.openai_model,
                    "instructions": system_prompt,
                    "input": f"Database context:\n{json.dumps(prompt_context, indent=2)}\n\nQuestion: {request.message}",
                    "reasoning": {"effort": "low"},
                    "text": {"verbosity": "low"},
                },
            )
            response.raise_for_status()
            answer = response_text(response.json())
            if not answer:
                raise ValueError("OpenAI returned an empty answer")
            model = settings.openai_model
            generated_by = "openai"
    except Exception:
        try:
            async with httpx.AsyncClient(timeout=settings.ollama_timeout_seconds) as client:
                response = await client.post(
                    f"{settings.ollama_base_url}/api/chat",
                    json={
                        "model": settings.ollama_model,
                        "stream": False,
                        "messages": [
                            {"role": "system", "content": system_prompt},
                            {"role": "user", "content": f"Database context:\n{json.dumps(prompt_context, indent=2)}\n\nQuestion: {request.message}"},
                        ],
                        "options": {"temperature": 0, "num_predict": 220},
                    },
                )
                response.raise_for_status()
                payload = response.json()
                answer = str(payload.get("message", {}).get("content", "")).strip()
                if not answer:
                    raise ValueError("Ollama returned an empty answer")
                model = str(payload.get("model") or settings.ollama_model)
                generated_by = "ollama"
        except Exception:
            answer = fallback_answer(context)
            model = "agripulse-local"
            generated_by = "fallback"

    result = evaluate_answer(request.message, answer, context, citations, generated_by)
    if result.should_fallback:
        answer = fallback_answer(context)
        model = "agripulse-local"
        generated_by = "fallback"
        result = evaluate_answer(request.message, answer, context, citations, generated_by)

    return ChatResponse(
        answer=answer,
        model=model,
        generated_by=generated_by,
        citations=citations,
        as_of=as_of,
        evaluation=result.evaluation,
    )
